### ISSUE TITLE
Refactor to zmq?

### ISSUE THREAD

**Issue by inf3rno** [inf3rno/dataflower#85]
I check the zmq documentation maybe I'll learn something worth to use.

http://zguide.zeromq.org/page:all

**Comment by inf3rno** [comment]
req-rep:
 - on the server you define a responder which you bind to a port
 - on the client you define a requester which you bind to the same port
 - the zmq protocol with the zmq context and the port is the contract between the client and the server

I can follow the same approach by dataflower:
 - on the upstream component I define an output port
 - on the downstream component I define an input port
 - the pipe between these ports is the contract

a simple echo server with this approach:

```
		init: function (){
			this.responder = ResponderFactory.newInstance({port: 1234});
			this.responder.on("message", this.notify.bind(this));
		},
		notify: function (message){
			this.responder.send(message);
		}
```

my approach if we have only 2 ports and respond immediately

```
		notify: function (message){
			this.send(message);
		}
```

my approach if we have multiple ports, so we have to use pull + push instead of req + rep

```
		notify: function (port){
			// this can contain conditional handling, correlation id definition, etc...
			// port = "input"
			this.handleRequest(this.ports[port].pull());
		},
		handleRequest: function (message){
			this.ports.output.push(message);
		}
```

I think this pull + push approach is better, because we can pull and push requests to the same ports. E.g.

```
		notify: function (){
			this.port.push(this.port.pull());
		}
```

```
		notify: function (port){
			port.push(port.pull());
		}
```

**Comment by inf3rno** [comment]
we could use neo4j cypher to connect components together:

```
var b = new o3.Class({
    init: function (){
        this.upstream = new df.DouplexStream({
            notify: this.upToDown.bind(this)
        });
        this.downstream = new df.DouplexStream({
            notify: this.downToUp.bind(this)
        });
    },
    upToDown: function (){
        this.downstream.push(this.upstream.pull());
    },
    downToUp: function (){
        this.upstream.push(this.downstream.pull());
    }
});

var pipeline =  df.cypher("(a.downstream) <-[ch1]-> (b.upstream)", {
    a: a,
    b: b
});
var ch1 = pipeline.ch1;
```

we can reuse the downstream -> upstream direction:

```
var pipeline =  df.cypher("(a) <-[ch1]-> (b) <-[ch2]-> (c) <-[ch3]-> (d)", {
    a: a,
    b: b,
    c: c,
    d: d,
    "^ch\d$":  function (left, right){
        return left.downstream.pipe(right.upstream);
    }
});
```

**Comment by inf3rno** [comment]
how to connect nodejs async functions:

```js
var NodeAdapter = o3.Class.extend({
    init: function (evaluate){
        this.stream = new df.DouplexStream({
            notify: this.notify.bind(this)
        });
        this.evaluate = evaluate;
    },
    notify: function (){
        var args = this.stream.pull();
        var cb = function (err, data){
            if (!err)
                this.stream.push(data);
        };
        this.evaluate(cb, args);
    }
});

var a = new NodeAdapter(function (cb, args){
    cb(null, args);
});
```

perfect... if we add correlation identifier to the cb, then we will be able to reply to a request.

```js
    notify: function (){
        var args = this.stream.pull();
        var messageId = this.stream.lastId();
        var cb = function (err, data){
            if (!err)
                this.stream.push({correlation: messageId}, data);
        };
        this.evaluate(cb, args);
    }
```

the channel has the responsibility to pair the request and the response using the correlation id.

**Comment by inf3rno** [comment]
using stack?

there are 2 simple options:
 - redefine the structure and invoke it with the params (this way we will use the stack of the js language
 - define a stack object and send it along with the messages (each message passing will clone the Object, so we will have a prototype chain at the end)

I think the first one is evident, we can use it manually if we want something simple.
The second one is less evident, it has to have a built in support. It is the russian doll approach, but we can extend it with correlation identifiers, etc... If we use infinite loop with this solution, then we will use up the memory very fast, because the too long prototype chain, so we have to count somehow the depth and drop the old stack if it is too deep.

**Comment by inf3rno** [comment]
what is the drawback of this approach?

nodejs streams contain drain and end events, while this doesn't.

 - the end is not necessary a problem, since we are not talking about chunks, just standalone messages. they can be chunks too, but we have to handle that with the application code, just as the end event. so for example by reading files, we have to use something like this:

```
StreamBegin(correlationId, range?)
StreamChunk(correlationId, range?, chunkData)
...
StreamEnd(correlationId, range?)
```

we can keep the message order, or if we don't then we can add a range param to restore the order when it is required.

**Comment by inf3rno** [comment]
zmq pull - push or ventillator - task - sink approach is similar to pump - filter - sink (pipes & filters), but they are not the same. pipes and filters describe that there is a directed data stream between these components, while zmq pull - push describes a possible implementation. 

the zmq task implementation is about pulling the messages when they have arrived. after we pulled them, we process them and push them forward. this is very similar to my pull - push approach. so I can describe this one too.

**Comment by inf3rno** [comment]
by zmq most of the problems is with starting the data flow. this is because the starting of the network is not regulated by a higher abstraction level code. it is just the same antipattern when we put initialization into a class instead of a factory or a di container. I think it is possible to do this properly by zmq as well, but I assume this requires skills and most zmq user does not have these skills. that's why it is an issue for example that publishers send messages before the subscribers connected to the port, etc...

**Comment by inf3rno** [comment]
there can be many different solutions here depending on duplicating the message, or sending it only to a single receiver or a group of receivers, etc... e.g. by pub-sub we duplicate it and send it to the receivers with the proper filter. by push/pull we send it to a single receiver. by req-rep we send it to a single receiver. and so on... the component code should contain this part. e.g. all receivers, switch, multiswitch, aggregation, etc...

**Comment by inf3rno** [comment]
by zmq the upper statements does not apply. yes the data flow automatically start, but you can stop it any time by its origin. so yes, we need a network builder code, which starts sending the messages, but this can be done with zmq. the problem is with people who does not use this tool in the way it's meant to use. (maybe my English is broken here)

**Comment by inf3rno** [comment]
> A TCP socket is an endpoint instance defined by an IP address and a port in the context of either a particular TCP connection or the listening state. A port is a virtualisation identifier defining a service endpoint (as distinct from a service instance endpoint aka session identifier). A TCP socket is not a connection, it is the endpoint of a specific connection. There can be concurrent connections to a service endpoint, because a connection is identified by both its local and remote endpoints, allowing traffic to be routed to a specific service instance. There can only be one listener socket for a given address/port combination.
http://stackoverflow.com/questions/152457/what-is-the-difference-between-a-port-and-a-socket

so

```
connection = socket1 <-> socket2
socket = new Endpoint()
Endpoint = ip address + port
```

according to the zmq docs, by most networks there are only a single connection by a single socket, while zmq behaves differently, by zmq many connections are bound to a single socket.

what are the proper terms by my solution, what words should I use?

```
var o = new X();
o.a = new Y();
o.a.on("notify", function Z (){
    var message = this.pull();
    this.push(message);
});
```
this is a simple echo component. what terms should we use?

```
var component = new Component();
component.socket = new Endpoint();
component.socket.on("notify", function process(){
    var message = this.pull();
    this.push(message);
});
```
hmm maybe this is okay, but what about

```
var address = new Address();
address.port = new Port();
address.port.on("notify", function listener(){
    var message = this.pull();
    this.push(message);
});
```

what is the problem here?
the Port is an abstraction, like an interface. it is something we bind our application to, so we should rather use it this way:

```
var address = new Address();
address.port = new Port();

var component = new Component();
component.socket = new Endpoint(address.port);
component.socket.on("notify", function process(){
    var message = this.pull();
    this.push(message);
});
```

so I think the first naming convention is much better. I am not sure about using the 1st or the 3rd solution. the first one is much simpler, while by the third one the components don't have to know anything about each other. it is similar to the mediator pattern. in most of the cases we don't need such decoupling and if we do so then we can use a mediator component. so I think the 1st approach is the best. I am still not sure about the naming convention, I have to work with it add more components, so I'll be able to decide whether keep these names or not.

```
var component = new Component();
component.socket = new Endpoint();
component.socket.on("notify", function process(){
    var message = this.pull();
    this.push(message);
});
```
e.g. queue could be a component which adds the messages to a list after push, and releases them only if they are pulled by the other socket of a connection. I am certain I have to work on these terms before defining an app specific vocab.

**Comment by inf3rno** [comment]
other definitions:
http://stackoverflow.com/questions/152457/what-is-the-difference-between-a-port-and-a-socket

> A port can be described as an internal address within a host that identifies a program or process. A socket can be described as a programming interface allowing a program to communicate with other programs or processes, on the internet, or locally. 

port = local address of a program
socket = interface to communicate between programs

nothing new.

> TCP-IP connections are bi-directional pathways connecting one address:port combination with another address:port combination. Therefore, whenever you open a connection from your local machine to a port on a remote server (say www.google.com:80), you are also associating a new port number on your machine with the connection, to allow the server to send things back to you, (e.g. 127.0.0.1:65234).

so port to port communication is directed and is unidirectional. so you cannot send data to 2 direction through a port. interesting. I don't think this is true. probably with HTTP this is true, but with websockets certainly not.

**Comment by inf3rno** [comment]
req-rep. in my push-pull model the req-rep is more like a connection type

```
->  |         |  ->  |         |  ->  |         |
    |    1    |      |    2    |      |    3    |
<-  |         |  <-  |         |  <-  |         |
```
each 3 components have 2 bidirectional connection here: an upstream and a downstream. they pull messages from the upstream and push messages to the downstream. when the downstream responds, they pull the response and push it back to the upstream. so both connections are req-rep connections. we can describe this structure as `... + req-rep(c1, c2) + req-rep(c2, c3) + ...`. if the processing is simple enough we can describe this in a general way. if it isn't then we have to use the pull and push methods of the upstream and downstream ports. the port is a connector. hmm we could then call it connector, that is meaningful name. I think I already used it.

so the components contain connectors and by binding connectors together we can make connections. 

```
componentA.connectorX.wire(componentB.connectorY)
```
by wiring these connectors together we can listen to the sender side. if it has message, it can notify the receiver side to pull the message. I think this can cause confusion. I mean connectorX.push(m) does not really push the message to connectorY, it just notifies the Y that there is message to receive. hmmm. zmq works a similar way, so I think this is an acceptable solution. I don't think we should rename these methods, there is no better vocab as far as I know. I don't like the connector and we don't need the component at least not be defining the connectors. connectors can be standalone.

**Comment by inf3rno** [comment]
I think I understand now these terms. I have a telephone analogy:
 - protocol: phoning technology standard
 - ip address + port: phone number
 - socket: phone device
 - connection: phone call
 - establishing connection: calling somebody
 - remote applications: people
 - sending messages: speech

So in my context:

```
var o = new X();
o.a = new Y();
o.a.on("notify", function Z (){
    var message = this.pull();
    this.push(message);
});
```

- X: Process/Application/Component
- o: Component instance
- Y: Socket
- a: socket //a specific socket the component uses, it should describe it's function
- Z: listener // a function which can react on messages, part of the Component if we follow the phone analogy

```
var echo = new Component({
    init: function (){
        this.socket = new Socket({
            message: this.listener.bind(this);
        });
    },
    listener: function (){
        var message = this.socket.receive();
        this.socket.send(message);
    }
});
```

ofc. we can use pull and push if we want, but I think receive and send is better and we already use those words.

**Comment by inf3rno** [comment]
what about things like Stack, Network, etc...? I am not sure I'll define them. I'll start with the common patterns: ReqRep, PubSub, PushPull, Queue, etc...

 - Actually we already have PushPull.
 - We can easily implement PubSub, we need a mediator component which receives messages from the publishers and sends them to the receivers. so it has to maintain the publisher and receiver list. the upgraded version could contain a message filter, so not every subscriber will get all of the messages. this is easy.
 - We can implement ReqRep too, what we need is message metadata, so we can add correlation id to messages and the receiver will be able to pair the requests with the responses. We can use the request itself instead of a correlation id, but it won't work if we are working with multiple processes or threads and we have to serialize the message.
 - By queue we can store the messages as long as they are not delivered to every receiver.

it is easy to implement these with the current system.

 - By implementing the Stack we need a unique Stack id and message metadata again. We can use the Stack object instead of the id if we want. The Stack instantiation and inheritance depends on the Network boundaries.
 - By defining Networks the NetworkBoundary could be a dedicated component for Stack instantiation or Stack destruction. This way we will be able to discover the full Component graph, and we don't have to put the inside of the Network into a closure.
 - The NetworkBoundary could do the dynamic construction of the inside of the Network if necessary. We could use a dedicated LazyLoader component or a the NetworkBoundary could have this responsibility as well.

I think this is solved. At least in theory. I'll read more about ZMQ and after that I'll implement this. I won't need the new o3 inheritance features. So when basic o3 features are stable, I can start this.

**Comment by inf3rno** [comment]
![dflower](https://cloud.githubusercontent.com/assets/408911/10435074/0ff686ca-711e-11e5-9f48-7edae743f9ef.png)
a simple connection between 2 components.
what is interesting that the connection is not directed, it is bidirectional. we can decide in which direction we use it. it will be harder to draw a directed component graph. if we need a directed connection we do the same, but we don't use one of the queues.

**Comment by inf3rno** [comment]
I cannot decide which one to use by pub/sub:

1.)

![1](https://cloud.githubusercontent.com/assets/408911/10435369/e06d1dfe-711f-11e5-8f52-dd7e3dd06c32.png)

socket.push(message) will deliver the same message through the connection

2.)

![2](https://cloud.githubusercontent.com/assets/408911/10435368/e06cb512-711f-11e5-811c-70ab9dfab0cc.png)

3.)

![3](https://cloud.githubusercontent.com/assets/408911/10435370/e07136b4-711f-11e5-8b37-0e92a45bb347.png)

**Comment by inf3rno** [comment]
I think the 1.) solution is the best and we don't need different connection types at least not for pub/sub.

**Comment by inf3rno** [comment]
What about switch or routing? By that if we have a single socket with many connections and we want to select which connections can receive the messages, we'd need a special socket type. We cannot have as many sockets as many routes we have. At least I don't think so, but I'll think about it later.

**Comment by inf3rno** [comment]
> To create a connection between two nodes, you use zmq_bind() in one node and zmq_connect() in the other. As a general rule of thumb, the node that does zmq_bind() is a "server", sitting on a well-known network address, and the node which does zmq_connect() is a "client", with unknown or arbitrary network addresses. Thus we say that we "bind a socket to an endpoint" and "connect a socket to an endpoint", the endpoint being that well-known network address.

This is different from df. By df we don't have to bind sockets to ports and call them endpoints. At least if we does not do networking then this is not necessary. Every socket has a dedicated object, so we can connect these objects together instead of connecting these objects to ports. Networking is ofc. somewhat different, but I'll work on a component which allows to move other components into threads and processes. This will require the serialization of the component and sending it through a socket connection and after that rebuilding it on the other side. This would be funny if we need something like a dynamic load balancer.

**Comment by inf3rno** [comment]
> One socket may have many outgoing and many incoming connections. There is no zmq_accept() method. When a socket is bound to an endpoint it automatically starts accepting connections. The network connection itself happens in the background, and ZeroMQ will automatically reconnect if the network connection is broken (e.g., if the peer disappears and then comes back). Your application code cannot work with these connections directly; they are encapsulated under the socket.

The one socket many connection part is okay. So endpoints are not sockets, sockets are bound to endpoints and after that they accept connections. Connections are encapsulated under the socket. Probably the endpoint is the socket identifier: ip + port. But I am not sure yet.

**Comment by inf3rno** [comment]
> Now, imagine we start the client before we start the server. In traditional networking, we get a big red Fail flag. But ZeroMQ lets us start and stop pieces arbitrarily. As soon as the client node does zmq_connect(), the connection exists and that node can start to write messages to the socket. At some stage (hopefully before messages queue up so much that they start to get discarded, or the client blocks), the server comes alive, does a zmq_bind(), and ZeroMQ starts to deliver messages.

This is interesting we have a queue without a real connection. It's like:
```js
var clientSocket = new Socket();
var connection = new Connection();
connection.add(clientSocket);
clientSocket.send(message1);
clientSocket.send(message2);
//...
var serverSocket = new Socket();
connection.add(serverSocket);
```
I really like this. I can do the same. So by a connection we can have a queue which will store the messages as long as somebody connects and pulls/receives the messages.

**Comment by inf3rno** [comment]
> Sockets have types. The socket type defines the semantics of the socket, its policies for routing messages inwards and outwards, queuing, etc. You can connect certain types of socket together, e.g., a publisher socket and a subscriber socket. Sockets work together in "messaging patterns". We'll look at this in more detail later.

I assume I was true about this. So it is better if we have multiple socket types and a single connection type...

**Comment by inf3rno** [comment]
so what will be the relation between connections and queues? should we add new queues by adding sockets to the connection and remove old queues by removing sockets from the connection? this would fix the directed connection problem. we won't need left and right and keep in mind which queue belongs to which socket. if we remove a socket, then the connection could store it's messages in a queue until they are processed. after that the queue could be removed. the only problem that sometimes the receiver socket needs to respond to the other socket. hmm that's not a problem, because it has it's dedicated queue, so it can send a message through that queue. ofc. nobody will receive it if nobody is using the connection, but that's another story. this is interesting, because we will be able to connect, send a message, disconnect, wait 5 mins for the response, connect again, receive the response and disconnect. I like this. it is similar to polling... ^^

**Comment by inf3rno** [comment]
nanomsg https://github.com/nickdesaulniers/node-nanomsg this is the zmq 2.0. the load balancer, worker etc... part could use this to send messages. actually all we need is a nanomsg adapter, which can do the job. and ofc we need a worker thread adapter too. it is somewhat faster to communicate with the thread then with the nanomsg through the network, that's why we need one adapter for threads and another for processes which will use nanomsg.

**Comment by inf3rno** [comment]
I think a connection should have multiple bound sockets and one receiver queue per connected socket. After disconnecting a socket the receiver queue is destroyed. By sending a message all receiver queue will be pushed. I was thinking on letting sockets to refuse messages, but I don't want to make this complicated. Let's stick with this implementation and check its limits on examples.
